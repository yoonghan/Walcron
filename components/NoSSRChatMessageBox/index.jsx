import * as React from "react";
import produce, {Draft} from "immer";
import dynamic from "next/dynamic";

const NoSSRChatFeed = dynamic(
  () => import('react-bell-chat').then(mod => mod.ChatFeed),
  { ssr: false }
)

export const NoSSRChatMessageBoxProps = {
  OTHERS: 1,
  SENDER: 2,
  SYSTEM: undefined
};

export class NoSSRChatMessageBox extends React.Component {

  constructor(props) {
    super(props);
    this.chatScroll = undefined;
    this.authors = this._initAuthors();
    this.authorsCount = this.authors.length;
    this.state = {
      messages: []
    }
  }

  _initAuthors = () => {
    return [
      {
        id: 1,
        name: 'Others',
        avatarName: "O",
        isTyping: true,
        lastSeenMessageId: 1,
        bgImageUrl: undefined
      },
      {
        id: 2,
        name: 'Sender',
        avatarName: "S",
        isTyping: false,
        lastSeenMessageId: 1,
        bgImageUrl: undefined
      }
    ];
  }

  addMessage = (senderId, message) => {
    this.setState(
      produce((draft) => {
        draft.messages.push({
          id: senderId,
          authorId: senderId,
          message: message,
          createdOn: new Date(),
          isSend: true
        });
      })
    );
  }

  _triggerFeed = (api) => {
    const chatScroll = this.chatScroll;
    if(api && api !== null) {
      if(chatScroll && chatScroll.scrollTo) {
        chatScroll.scrollTo(0, chatScroll.scrollHeight);
      }
      else {
        this.chatScroll = document.getElementsByClassName("react-bell-chat__chat-scroll-area")[0];
      }
    }

  }

  render() {
    const {messages} = this.state;
    return (
      <div className={"container"}>
        <NoSSRChatFeed
          messages={messages}
          authors={this.authors}
          yourAuthorId={this.authorsCount}
          showRecipientAvatar={true}
          minHeight = {400}
          maxHeight = {400}
          ref={this._triggerFeed}
        />
        <style jsx>{`
          .container {
            background-color: #FFFFFF;
          }
        `}
        </style>
      </div>
    )
  }
}

export default NoSSRChatMessageBox;
