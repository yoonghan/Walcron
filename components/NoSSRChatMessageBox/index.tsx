import * as React from "react";
import produce, {Draft} from "immer";
import dynamic from "next/dynamic";
import {Author, Message, ChatFeedApi} from "react-bell-chat";

interface NoSSRChatMessageBoxProps {
}

interface NoSSRChatMessageBoxState {
  messages: Array<Message>;
}

const NoSSRChatFeed = dynamic(
  () => import('react-bell-chat').then(mod => mod.ChatFeed),
  { ssr: false }
)

export class NoSSRChatMessageBox extends React.Component<NoSSRChatMessageBoxProps, NoSSRChatMessageBoxState> {
  private authors:Array<Author>;
  private authorsCount:number;
  public static OTHERS = 1;
  public static SENDER = 2;
  public static SYSTEM = undefined;
  private chatScroll:any;

  constructor(props:NoSSRChatMessageBoxProps) {
    super(props);
    this.authors = this._initAuthors();
    this.authorsCount = this.authors.length;
    this.state = {
      messages: []
    }
  }

  _initAuthors = () => {
    return [
      {
        id: 1,
        name: 'Others',
        avatarName: "O",
        isTyping: true,
        lastSeenMessageId: 1,
        bgImageUrl: undefined
      },
      {
        id: 2,
        name: 'Sender',
        avatarName: "S",
        isTyping: false,
        lastSeenMessageId: 1,
        bgImageUrl: undefined
      }
    ];
  }

  addMessage = (senderId:number|undefined, message:string) => {
    this.setState(
      produce((draft: Draft<NoSSRChatMessageBoxState>) => {
        draft.messages.push({
          id: senderId,
          authorId: senderId,
          message: message,
          createdOn: new Date(),
          isSend: true
        });
      })
    );
  }

  _triggerFeed = (api: ChatFeedApi) => {
    const chatScroll = this.chatScroll;
    if(api && api !== null) {
      if(chatScroll && chatScroll.scrollTo) {
        (chatScroll as any).scrollTo(0, chatScroll.scrollHeight);
      }
      else {
        this.chatScroll = document.getElementsByClassName("react-bell-chat__chat-scroll-area")[0];
      }
    }

  }

  render() {
    const {messages} = this.state;
    return (
      <div className={"container"}>
        <NoSSRChatFeed
          messages={messages}
          authors={this.authors}
          yourAuthorId={this.authorsCount}
          showRecipientAvatar={true}
          minHeight = {400}
          maxHeight = {400}
          ref={this._triggerFeed}
        />
        <style jsx>{`
          .container {
            background-color: #FFFFFF;
          }
        `}
        </style>
      </div>
    )
  }
}

export default NoSSRChatMessageBox;
