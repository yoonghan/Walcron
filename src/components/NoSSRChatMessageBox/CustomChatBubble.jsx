import * as React from "react";
import dynamic from "next/dynamic";

const NoSSRChatBubble = dynamic(
  () => import('react-bell-chat').then(mod => mod.ChatBubble),
  { ssr: false }
)

var linkRegex = new RegExp("^(http|https)://");

const CustomChatBubble = (props) => {

  const _renderLink = () => {
    const youAreAuthor = props.message.authorId === props.yourAuthorId;

    return (
      <div className={"layer-1"}>
        <div className={youAreAuthor ? "author-bubble": "sender-bubble"}>
          <a href={props.message.message} target="walcron_tab">
            Shared file - [Click to view]
          </a>
          <span className={"time"}>
            {props.message.createdOn.toLocaleTimeString([], {
              hour: '2-digit',
              minute: '2-digit',
              hour12: true
            })}
          </span>
        </div>
        <style jsx>{`
          .layer-1 {
            position: relative;
            margin-bottom: 2px;
            display: flex;
          }
          .author-bubble {
            position: relative;
            background-color: rgb(0, 132, 255);
            border-radius: 7px 0px 7px 7px;
            margin-right: 0px;
            margin-left: auto;
            max-width: 80%;
            padding: 8px 14px 14px;
          }
          .sender-bubble {
            position: relative;
            background-color: rgb(204, 204, 204);
            border-radius: 0px 7px 7px 0px;
            margin-right: auto;
            margin-left: 0px;
            max-width: 80%;
            padding: 8px 14px 14px;
          }
          .time {
            position: absolute;
            right: 14px;
            bottom: 2px;
            font-size: 9px;
            color: rgba(255, 255, 255, 0.55);
          }
        `}</style>
      </div>
    )
  }

  const _renderChat = () => {
    const isLink = linkRegex.test(props.message.message);
    if(isLink) {
      return _renderLink();
    }
    else {
      return <NoSSRChatBubble {...props}/>;
    }
  }

  return _renderChat();
}

export default CustomChatBubble;
