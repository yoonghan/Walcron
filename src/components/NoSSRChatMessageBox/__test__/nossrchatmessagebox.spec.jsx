import * as React from 'react';
import { shallow } from 'enzyme';
import NoSSRChatMessageBox from "../index";

describe('NoSSRChatMessageBox', () => {
  it('renders normally', () => {
    const wrapper = shallow(
      <NoSSRChatMessageBox/>
    );
    expect(wrapper.debug()).toMatchSnapshot();
  })
});
