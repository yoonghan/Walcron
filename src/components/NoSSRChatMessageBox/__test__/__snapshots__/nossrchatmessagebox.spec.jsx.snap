// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoSSRChatMessageBox renders normally 1`] = `
"<div className=\\"jsx-1657555887 container\\">
  <Dynamic customChatBubble={[Function: CustomChatBubble]} messages={{...}} authors={{...}} yourAuthorId={2} showRecipientAvatar={true} minHeight={0} maxHeight={0} />
  <JSXStyle id=\\"1657555887\\">
    .container.jsx-1657555887{background-color:#FFFFFF;}
  </JSXStyle>
</div>"
`;
